libs/elasticsearch-nio/src/main/java/org/elasticsearch/nio/NioSocketChannel.java

```diff
@@ -35,9 +32,8 @@
     private final InetSocketAddress remoteAddress;
     private final CompletableFuture<Void> connectContext = new CompletableFuture<>();
     private final SocketSelector socketSelector;
-    private final AtomicBoolean contextsSet = new AtomicBoolean(false);
-    private ChannelContext context;
-    private BiConsumer<NioSocketChannel, Exception> exceptionContext;
+    private final AtomicBoolean contextSet = new AtomicBoolean(false);
+    private SocketChannelContext context;
```

Comment:
why do we have a context here again? the superclass has it aready? Can't we use a getter that has the right type or a generic?

Reply:
Is there any change we can untangle the cyclic dependencies here and pass the context in as a ctor argument? or can we pass it as method arguments instead and don't have it as a member?

Reply:
I removed the duplicate contexts. Now each stances (`NioServerSocketChannel` and `NioSocketChannel`) has their own. I could store it in the super class parameterized. But `AbstractNioChannel` is already parameterized once and it would get pretty big.

Probably not possible to change the cyclic relationship right now. Context needs the channel. So it cannot be a ctor argument without allowing `this` to escape. Channel needs context as a field, as it is the hook into lower level selector operations. But going forward I have some ideas about how to continue to reduce this relationship.

Reply:
cool lets move on with this.